File: backend/services/auth-go/internal/service/auth_service.go
package service

import (
	"auth-go/internal/dto"
	"auth-go/internal/model"
	"auth-go/internal/repository"
	"auth-go/internal/util"
	"context"
	"errors"
	"time"

	"github.com/redis/go-redis/v9"
	"golang.org/x/crypto/bcrypt"
)

type AuthService interface {
	Register(req *dto.RegisterRequest) (*dto.AuthResponse, error)
	Login(req *dto.LoginRequest) (*dto.AuthResponse, error)
	Logout(ctx context.Context, authHeader string) error
	RefreshToken(refreshToken string) (*dto.AuthResponse, error)
	GetCurrentUser(email string) (*dto.UserProfile, error)
}

type authService struct {
	userRepo    repository.UserRepository
	sessionRepo repository.SessionRepository
	jwtUtil     *util.JwtUtil
	redis       *redis.Client
}

func NewAuthService(userRepo repository.UserRepository, sessionRepo repository.SessionRepository, jwtUtil *util.JwtUtil, redis *redis.Client) AuthService {
	return &authService{
		userRepo:    userRepo,
		sessionRepo: sessionRepo,
		jwtUtil:     jwtUtil,
		redis:       redis,
	}
}

func (s *authService) Register(req *dto.RegisterRequest) (*dto.AuthResponse, error) {
	exists, err := s.userRepo.ExistsByEmail(req.Email)
	if err != nil {
		return nil, err
	}
	if exists {
		return nil, errors.New("email already exists")
	}

	hashedPassword, err := bcrypt.GenerateFromPassword([]byte(req.Password), bcrypt.DefaultCost)
	if err != nil {
		return nil, err
	}

	user := &model.User{
		Email:        req.Email,
		PasswordHash: string(hashedPassword),
		Phone:        req.Phone,
		UserType:     req.UserType,
		IsActive:     true,
	}

	if err := s.userRepo.Save(user); err != nil {
		return nil, err
	}

	accessToken, err := s.jwtUtil.GenerateToken(user.Email, user.UserType, user.ID)
	if err != nil {
		return nil, err
	}

	refreshToken, err := s.jwtUtil.GenerateRefreshToken(user.Email)
	if err != nil {
		return nil, err
	}

	if err := s.saveSession(user, refreshToken, ""); err != nil {
		return nil, err
	}

	return &dto.AuthResponse{
		AccessToken:  accessToken,
		RefreshToken: refreshToken,
	}, nil
}

func (s *authService) Login(req *dto.LoginRequest) (*dto.AuthResponse, error) {
	user, err := s.userRepo.FindByEmail(req.Email)
	if err != nil {
		return nil, errors.New("invalid credentials")
	}

	if err := bcrypt.CompareHashAndPassword([]byte(user.PasswordHash), []byte(req.Password)); err != nil {
		return nil, errors.New("invalid credentials")
	}

	accessToken, err := s.jwtUtil.GenerateToken(user.Email, user.UserType, user.ID)
	if err != nil {
		return nil, err
	}

	refreshToken, err := s.jwtUtil.GenerateRefreshToken(user.Email)
	if err != nil {
		return nil, err
	}

	if err := s.saveSession(user, refreshToken, req.DeviceInfo); err != nil {
		return nil, err
	}

	return &dto.AuthResponse{
		AccessToken:  accessToken,
		RefreshToken: refreshToken,
	}, nil
}

func (s *authService) Logout(ctx context.Context, authHeader string) error {
	if authHeader == "" || len(authHeader) < 7 {
		return nil
	}
	token := authHeader[7:]

	expiry, err := s.jwtUtil.ExtractExpiration(token)
	if err != nil {
		return nil
	}

	duration := time.Until(expiry)
	if duration > 0 {
		return s.redis.Set(ctx, "blacklist:"+token, "true", duration).Err()
	}
	return nil
}

func (s *authService) RefreshToken(refreshToken string) (*dto.AuthResponse, error) {
	session, err := s.sessionRepo.FindByToken(refreshToken)
	if err != nil {
		return nil, errors.New("invalid refresh token")
	}

	if session.ExpiresAt.Before(time.Now()) {
		s.sessionRepo.Delete(session)
		return nil, errors.New("refresh token expired")
	}

	user := &session.User
	newAccessToken, err := s.jwtUtil.GenerateToken(user.Email, user.UserType, user.ID)
	if err != nil {
		return nil, err
	}

	return &dto.AuthResponse{
		AccessToken:  newAccessToken,
		RefreshToken: refreshToken,
	}, nil
}

func (s *authService) GetCurrentUser(email string) (*dto.UserProfile, error) {
	user, err := s.userRepo.FindByEmail(email)
	if err != nil {
		return nil, errors.New("user not found")
	}

	return &dto.UserProfile{
		ID:       user.ID,
		Email:    user.Email,
		Phone:    user.Phone,
		UserType: user.UserType,
	}, nil
}

func (s *authService) saveSession(user *model.User, token string, deviceInfo string) error {
	expiresAt, err := s.jwtUtil.ExtractExpiration(token)
	if err != nil {
		return err
	}

	session := &model.Session{
		UserID:     user.ID,
		Token:      token,
		ExpiresAt:  expiresAt,
		DeviceInfo: deviceInfo,
	}

	return s.sessionRepo.Save(session)
}

File: backend/services/auth/src/main/java/org/auth/service/AuthService.java
package org.auth.service;

import lombok.RequiredArgsConstructor;
import org.auth.dto.AuthResponse;
import org.auth.dto.LoginRequest;
import org.auth.dto.RegisterRequest;
import org.auth.dto.UserProfile;
import org.auth.model.Session;
import org.auth.model.User;
import org.auth.repository.SessionRepository;
import org.auth.repository.UserRepository;
import org.auth.util.JwtUtil;
import org.auth.exception.InvalidTokenException;
import org.auth.exception.ResourceNotFoundException;
import org.auth.exception.UserAlreadyExistsException;
import org.springframework.data.redis.core.StringRedisTemplate;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.time.ZoneId;
import java.util.Objects;
import java.util.concurrent.TimeUnit;

@Service
@RequiredArgsConstructor
public class AuthService {

    private final UserRepository userRepository;
    private final SessionRepository sessionRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtUtil jwtUtil;
    private final AuthenticationManager authenticationManager;
    private final StringRedisTemplate redisTemplate;

    @Transactional
    public AuthResponse register(RegisterRequest request) {
        if (userRepository.existsByEmail(request.email())) {
            throw new UserAlreadyExistsException("Email already exists");
        }

        User user = User.builder()
                .email(request.email())
                .password(passwordEncoder.encode(request.password()))
                .phone(request.phone())
                .userType(request.userType())
                .isActive(true)
                .build();

        userRepository.save(user);

        String accessToken = jwtUtil.generateToken(user.getEmail(), user.getUserType(), user.getId());
        String refreshToken = jwtUtil.generateRefreshToken(user.getEmail());

        saveSession(user, refreshToken, null);

        return AuthResponse.builder()
                .accessToken(accessToken)
                .refreshToken(refreshToken)
                .build();
    }

    @Transactional
    public AuthResponse login(LoginRequest request) {
        authenticationManager.authenticate(
                new UsernamePasswordAuthenticationToken(request.email(), request.password())
        );

        User user = userRepository.findByEmail(request.email())
                .orElseThrow(() -> new ResourceNotFoundException("User not found"));

        String accessToken = jwtUtil.generateToken(user.getEmail(), user.getUserType(), user.getId());
        String refreshToken = jwtUtil.generateRefreshToken(user.getEmail());

        saveSession(user, refreshToken, request.deviceInfo());

        return AuthResponse.builder()
                .accessToken(accessToken)
                .refreshToken(refreshToken)
                .build();
    }

    @Transactional
    public void logout(String authHeader) {
        if (authHeader != null && authHeader.startsWith("Bearer ")) {
            String token = authHeader.substring(7);
            long expiry = jwtUtil.extractExpiration(token).getTime() - System.currentTimeMillis();
            if (expiry > 0) {
                redisTemplate.opsForValue().set("blacklist:" + token, "true", expiry, TimeUnit.MILLISECONDS);
            }
        }
    }

    @Transactional
    public AuthResponse refreshToken(String refreshToken) {
        Session session = sessionRepository.findByToken(refreshToken)
                .orElseThrow(() -> new InvalidTokenException("Invalid refresh token"));

        if (session.getExpiresAt().isBefore(LocalDateTime.now())) {
            sessionRepository.delete(session);
            throw new InvalidTokenException("Refresh token expired");
        }

        User user = session.getUser();
        String newAccessToken = jwtUtil.generateToken(user.getEmail(), user.getUserType(), user.getId());

        return AuthResponse.builder()
                .accessToken(newAccessToken)
                .refreshToken(refreshToken)
                .build();
    }

    public UserProfile getCurrentUser() {
        String email = Objects.requireNonNull(SecurityContextHolder.getContext().getAuthentication()).getName();
        User user = userRepository.findByEmail(email)
                .orElseThrow(() -> new ResourceNotFoundException("User not found"));

        return UserProfile.builder()
                .id(user.getId())
                .email(user.getEmail())
                .phone(user.getPhone())
                .userType(user.getUserType())
                .build();
    }

    private void saveSession(User user, String token, String deviceInfo) {
        LocalDateTime expiresAt = jwtUtil.extractExpiration(token)
                .toInstant().atZone(ZoneId.systemDefault()).toLocalDateTime();

        Session session = Session.builder()
                .user(user)
                .token(token)
                .expiresAt(expiresAt)
                .deviceInfo(deviceInfo)
                .build();

        sessionRepository.save(session);
    }
}
